sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -2833,14 +2812,9 @@ pi_result piextQueueGetNativeHandle(pi_queue Queue,
   auto ZeQueue = pi_cast<ze_command_queue_handle_t *>(NativeHandle);
 
   // Extract a Level Zero compute queue handle from the given PI queue
+  auto &QueueGroup = Queue->getQueueGroup(false /*compute*/);
   uint32_t QueueGroupOrdinalUnused;
-  auto TID = std::this_thread::get_id();
-  auto &InitialGroup = Queue->ComputeQueueGroupsByTID.begin()->second;
-  const auto &Result =
-      Queue->ComputeQueueGroupsByTID.insert({TID, InitialGroup});
```

Comment:
This code was doing wrong it used the same `ZeQueues` for the new entry so it is then was destroyed multiple times causing a segfault. Now, without per-thread setting we don't even create multiple entries, and we do per-thread then the new entry is properly initialized here: https://github.com/intel/llvm/pull/8896/files#diff-1f19baf7cc76d94bf4a1819c81e7b34292fb3ce215cd50873e8c9236cc73601cR501

Reply:
This feature was for immediate command lists only. So the zeQueues should have been unallocated. Have you extended it for command queues also?

Reply:
So now the _vector_ of ZeQueues is now separate for each entry?

Reply:
See here, new entries are initialized with null: https://github.com/intel/llvm/pull/8896/files#diff-1f19baf7cc76d94bf4a1819c81e7b34292fb3ce215cd50873e8c9236cc73601cR501

Reply:
It is still for immediate command-lists only, but it is safer to initialize `ZeQueues` to null for safety

Reply:
OK, so at Map destruction, as each entry is being destroyed, the ZeQueues vector which has a non-zero size, each entry would also be destroyed. Was there an expliciit call to zecommandqueue destroy? And where did that call come from?

Reply:
The restructuring looks good. Just trying to understand the failure.

Reply:
All `ZeQueues` from all entries are being destrooyed: https://github.com/intel/llvm/blob/24ae0153e8e8a492c6a4feed5eeeb2fa1e26204b/sycl/plugins/level_zero/pi_level_zero.cpp#L2761

And there was a bug that multiple entries hold the same `ZeQueues`

Reply:
But that value should have been a null, and the zecommandqueuedestroy should not have been called. But it did get called, which means that the vector of zequeus has to be explicitly set to nullptrs. OK.

Reply:
it was nullptr but then a call to`getZeQueue` created a non-null one: https://github.com/intel/llvm/pull/8896/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985L2843

That non-null value was then copied to yet a new entry created for another thread.

Reply:
The perThread and getNative don't mix well. Say you have a queue that uses multiple cmd lists, one per thread. Someone calls getNative, makes his own queue from it and assumes that a sync on that queue is also a sync on the original queue. But it isn't because only one cmd list will be synch'ed.

Reply:
Right, but the failure was even without per-thread requested. We still created multiple redundant entries, which additionally held the same queues. I don't think that syncing a queue returned by a get_native() gives any guarantees about the entire SYCL queue being synced.

Reply:
That insert was supposed to be "insert if not found".